src/main/java/com/ibm/disni/util/MemoryAllocation.java

```diff
@@ -21,67 +21,98 @@
 	}
-	
-	private MemoryAllocation() {
-		table = new ConcurrentHashMap<String, MemBuf>();
-	}
-	
-	public MemBuf allocate(int size, MemType type, String classname) {
-		if (type == MemType.DIRECT) {
-			String key = generateKey(size, type, classname);
-			MemBuf buf = table.remove(key);
-			if (buf != null) {
-				buf.getBuffer().clear();
-				return buf;
+
+	private class AllocatorStack {
+		private final ConcurrentLinkedDeque<MemBuf> stack = new ConcurrentLinkedDeque<>();
+		private final int size;
+
+
+		private AllocatorStack(int size) {
+			this.size = size;
+		}
+
+
+		private MemBuf get() {
+			MemBuf buffer = stack.pollFirst();
+			if (buffer == null) {
+				logger.debug("Allocating buffer of size {}" , size);
+				ByteBuffer buff = ByteBuffer.allocateDirect(size);
+				buff.order(ByteOrder.nativeOrder());
+				return new MemBuf(buff);
+			} else {
+				logger.debug("Reusing buffer of size {}", size);
+				return buffer;
+			}
+		}
+
+		private void put(MemBuf buffer) {
+			buffer.getBuffer().clear();
+			stack.addLast(buffer);
+		}
+
+		public void preallocate(int numBuffers){
+			logger.debug("Pre allocating {} buffer of size {} B", numBuffers, size);
+			for (int i = 0; i < numBuffers; i++) {
+				ByteBuffer buff = ByteBuffer.allocateDirect(size);
+				buff.order(ByteOrder.nativeOrder());
+				put(new MemBuf(buff));
 			}
 		}
-		return _allocate(size, type, classname);
 	}
-	
-
-	public void free(MemBuf memBuf) {
-		if (memBuf.getType() == MemType.DIRECT) {
-			String key = generateKey(memBuf.getBuffer().capacity(),
-					memBuf.getType(), memBuf.getClassname());
-			table.put(key, memBuf);
-		} 
+
+
+	public MemBuf allocate(int size){
+
+		if (size < MIN_BLOCK_SIZE) {
+			size = MIN_BLOCK_SIZE;
+		} else {
+			size--;
+			size |= size >> 1;
+			size |= size >> 2;
+			size |= size >> 4;
+			size |= size >> 8;
+			size |= size >> 16;
+			size++;
+		}
+		return getOrCreateAllocatorStack(size).get();
 	}
 
-	private String generateKey(int size, MemType type, String classname) {
-		String key = classname + ":" + size + ":" + type.ordinal();
-		return key;
+	private AllocatorStack getOrCreateAllocatorStack(int length) {
+		AllocatorStack allocatorStack = allocStackMap.get(length);
+		if (allocatorStack == null) {
+			allocStackMap.putIfAbsent(length, new AllocatorStack(length));
+			allocatorStack = allocStackMap.get(length);
+		}
+
+		return allocatorStack;
 	}
 
-	private MemBuf _allocate(int size, MemType type, String classname) {
-		if (type == MemType.INDIRECT) {
-			ByteBuffer buffer = ByteBuffer.allocate(size);
-			buffer.order(ByteOrder.nativeOrder());
-			long address = 0;
-			return new MemBuf(MemType.INDIRECT, buffer, address, this,
-					classname);
-		} else {
-			ByteBuffer buffer = ByteBuffer.allocateDirect(size);
-			buffer.order(ByteOrder.nativeOrder());
-			long address = MemoryUtils.getAddress(buffer);
-			return new MemBuf(MemType.DIRECT, buffer, address, this, classname);
+	public void put(MemBuf buf){
```

Comment:
You should put a cap on the maximum idle buffers in the pool, you can make it configurable and set some default. If the cap is met, then it should free the least used pool.